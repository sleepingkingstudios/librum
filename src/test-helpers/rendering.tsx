import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
