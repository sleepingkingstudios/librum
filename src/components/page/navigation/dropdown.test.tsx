import * as React from 'react';

import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import {
  Dropdown,
  DropdownElementProps,
} from './dropdown';
import { render } from '@test-helpers/rendering';

describe('<Dropdown>', () => {
  const defaultProps: DropdownElementProps = {
    label: 'Rockets',
    items: [
      {
        label: 'Engines',
        url: '/rockets/engines',
      },
      {
        label: 'Fuel Tanks',
        url: '/rockets/fuel-tanks',
      },
    ],
    open: null,
    setOpen: jest.fn(),
  };

  describe('when all dropdowns are closed', () => {
    const closedProps: DropdownElementProps = {
      ...defaultProps,
      open: null,
    };

    it('should render the toggle', () => {
      const { getByRole } = render(<Dropdown {...closedProps} />);

      const button = getByRole('button', { name: defaultProps.label });

      expect(button).toBeVisible();
    });

    it('should call setOpen', async () => {
      const updateState = jest.fn();
      const { getByRole } =
        render(<Dropdown {...closedProps} setOpen={updateState} />);

      const button = getByRole('button', { name: defaultProps.label });

      await userEvent.click(button);

      expect(updateState).toHaveBeenCalledWith(defaultProps.label);
    });

    it('should not render dropdown items', () => {
      const { queryByText } = render(<Dropdown {...closedProps} />);

      defaultProps.items.forEach(
        ({ label }) => expect(queryByText(label)).toBeNull()
      )
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(<Dropdown {...closedProps} />);

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when another dropdown is open', () => {
    const closedProps: DropdownElementProps = {
      ...defaultProps,
      open: '/launch-sites',
    };

    it('should render the toggle', () => {
      const { getByRole } = render(<Dropdown {...closedProps} />);

      const button = getByRole('button', { name: defaultProps.label });

      expect(button).toBeVisible();
    });

    it('should call setOpen', async () => {
      const updateState = jest.fn();
      const { getByRole } =
        render(<Dropdown {...closedProps} setOpen={updateState} />);

      const button = getByRole('button', { name: defaultProps.label });

      await userEvent.click(button);

      expect(updateState).toHaveBeenCalledWith(defaultProps.label);
    });

    it('should not render dropdown items', () => {
      const { queryByText } = render(<Dropdown {...closedProps} />);

      defaultProps.items.forEach(
        ({ label }) => expect(queryByText(label)).toBeNull()
      )
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(<Dropdown {...closedProps} />);

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when the dropdown is open', () => {
    const openProps: DropdownElementProps = {
      ...defaultProps,
      open: defaultProps.label,
    };

    it('should render the toggle', () => {
      const { getByRole } =
        render(<Dropdown {...openProps} />, { router: true });

      const button = getByRole('button', { name: defaultProps.label });

      expect(button).toBeVisible();
    });

    it('should call setOpen', async () => {
      const updateState = jest.fn();
      const { getByRole } = render(
        <Dropdown {...openProps} setOpen={updateState} />,
        { router: true }
      );

      const button = getByRole('button', { name: defaultProps.label });

      await userEvent.click(button);

      expect(updateState).toHaveBeenCalledWith(null);
    });

    it('should render the dropdown items', () => {
      const { queryByRole } =
        render(<Dropdown {...openProps} />, { router: true });

      defaultProps.items.forEach(
        ({ label }) => {
          expect(queryByRole('link', { name: label })).toBeVisible();
        }
      )
    });

    it('should match the snapshot', () => {
      const { asFragment } =
        render(<Dropdown {...openProps} />, { router: true });

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
