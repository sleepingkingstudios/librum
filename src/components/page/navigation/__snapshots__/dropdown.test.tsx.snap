// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown> when all dropdowns are closed should match the snapshot 1`] = `
<DocumentFragment>
  <li
    class="dropdown-container"
  >
    <div
      class="dropdown"
    >
      <button
        class="dropdown-label"
      >
        Rockets
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-caret-up fa-xs ml-2"
          data-icon="caret-up"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M182.6 137.4c-12.5-12.5-32.8-12.5-45.3 0l-128 128c-9.2 9.2-11.9 22.9-6.9 34.9s16.6 19.8 29.6 19.8H288c12.9 0 24.6-7.8 29.6-19.8s2.2-25.7-6.9-34.9l-128-128z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
  </li>
</DocumentFragment>
`;

exports[`<Dropdown> when another dropdown is open should match the snapshot 1`] = `
<DocumentFragment>
  <li
    class="dropdown-container"
  >
    <div
      class="dropdown"
    >
      <button
        class="dropdown-label"
      >
        Rockets
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-caret-up fa-xs ml-2"
          data-icon="caret-up"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M182.6 137.4c-12.5-12.5-32.8-12.5-45.3 0l-128 128c-9.2 9.2-11.9 22.9-6.9 34.9s16.6 19.8 29.6 19.8H288c12.9 0 24.6-7.8 29.6-19.8s2.2-25.7-6.9-34.9l-128-128z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
  </li>
</DocumentFragment>
`;

exports[`<Dropdown> when the dropdown is open should match the snapshot 1`] = `
<DocumentFragment>
  <li
    class="dropdown-container"
  >
    <div
      class="dropdown"
    >
      <button
        class="dropdown-label"
      >
        Rockets
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-caret-down fa-xs ml-2"
          data-icon="caret-down"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M137.4 374.6c12.5 12.5 32.8 12.5 45.3 0l128-128c9.2-9.2 11.9-22.9 6.9-34.9s-16.6-19.8-29.6-19.8L32 192c-12.9 0-24.6 7.8-29.6 19.8s-2.2 25.7 6.9 34.9l128 128z"
            fill="currentColor"
          />
        </svg>
      </button>
      <ul>
        <li
          class="dropdown-item"
        >
          <a
            class="dropdown-item-link"
            href="/rockets/engines"
          >
            Engines
          </a>
        </li>
        <li
          class="dropdown-item"
        >
          <a
            class="dropdown-item-link"
            href="/rockets/fuel-tanks"
          >
            Fuel Tanks
          </a>
        </li>
      </ul>
    </div>
  </li>
</DocumentFragment>
`;
